contracts/ERC721A.sol

```diff
@@ -580,15 +580,36 @@ contract ERC721A is IERC721A {
     }
 
     /**
-     * @dev Zeroes out _tokenApprovals[tokenId]
+     * @dev Returns the storage slot and value for the approved address of `tokenId`.
      */
-    function _removeTokenApproval(uint256 tokenId) private {
-        mapping(uint256 => address) storage tokenApprovalPtr = _tokenApprovals;
+    function _getApprovedAddress(
+        uint256 tokenId
+    ) private view returns (uint256 approvedAddressSlot, address approvedAddress) {
+        mapping(uint256 => address) storage tokenApprovalsPtr = _tokenApprovals;
+        // The following is equivalent to `approvedAddress[tokenId]`.
         assembly {
+            // Compute the slot.    
             mstore(0x00, tokenId)
-            mstore(0x20, tokenApprovalPtr.slot)
-            let hash := keccak256(0, 0x40)
-            sstore(hash, 0)
+            mstore(0x20, tokenApprovalsPtr.slot)
+            approvedAddressSlot := keccak256(0x00, 0x40)
+            // Load the slot's value from storage.
+            approvedAddress := sload(approvedAddressSlot)
+        }
+    }
+
+    /**
+     * @dev Returns whether the `approvedAddress` is equals to `from` or `msgSender`.
+     */
+    function _isOwnerOrApproved(
+        address approvedAddress,
+        address from, 
+        address msgSender
```

Comment:
Do we need to make this an input param, or can we just call `msgSenderERC721A()` directly? Seems like the goal of this is to always validate against msgSender, so there's no point in passing it in as an arg

Reply:
Yes, `_msgSenderERC721A()` may be overridden for GSN contracts.